internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -2153,4 +2203,44 @@ public String getDestination() {
   public PubSubProducerAdapter getProducerAdapter() {
     return this.producerAdapter;
   }
+
+  /**
+   * {@link VeniceComplexPartitioner} offers a more sophisticated getPartitionId API. It takes value as a parameter, and
+   * it could return a single, multiple or no partition(s). The delete behavior is also undefined, so we will ignore
+   * deletes via key for now. Proper delete can be performed via re-push or new version push.
+   */
+  public CompletableFuture<PubSubProduceResult> writeWithComplexPartitioner(
```

Comment:
What does the returned CF mean? That all partitions have been successfully produced into? What if there is a partial failure? Should this be a collection of results instead?

Reply:
Correct, the current implementation returns a CF that's completed if all partitions have been successfully produced into. If there is a partial failure then the returned CF is completed exceptionally and there will be no way for caller to know which partition succeeded and which partitions failed.

Reply:
Ok well I guess that is fine, BUT... we should explain that in the JavaDoc, and IMHO it doesn't make sense to return a `PubSubProduceResult` which inherently represents just one partition. It can be a `CompletableFuture<Void>` I guess, or else return some new type which makes sense in that context, if there is anything we really need in there...